### PR TITLE
Moved precomputeStressFromQInterpolated back to c++ from yateto

### PR DESCRIPTION
**Pull request by ravil-mobile**

### CODE REVIEW

**Comment by ravil-mobile** [comment]
Seems like with `clang@12 -O3 -DNDEBUG   -mavx2 -mfma` everything is ok. gcc is also ok.
```
precomputeStressFromQInterpolated(unsigned int): # @precomputeStressFromQInterpolated(unsigned int)
        push    r15
        push    r14
        push    r13
        push    r12
        push    rbx
        mov     r14, rdi
        mov     r15, qword ptr [rip + qInterpolatedPlus]
        mov     r12d, esi
        mov     r13, qword ptr [rip + qInterpolatedMinus]
        xor     ebx, ebx
        mov     edx, 8064
        xor     esi, esi
        call    memset@PLT
        imul    rcx, r12, 24192
        lea     rax, [rcx + r13]
        add     rax, 3584
        add     rcx, r15
        add     rcx, 3584
        mov     rdx, r14
        add     rdx, 5376
        vbroadcastsd    ymm0, qword ptr [rip + .LCPI0_0] # ymm0 = [3.2000000000000002E+0,3.2000000000000002E+0,3.2000000000000002E+0,3.2000000000000002E+0]
        vbroadcastsd    ymm1, qword ptr [rip + .LCPI0_1] # ymm1 = [3.1000000000000001E+0,3.1000000000000001E+0,3.1000000000000001E+0,3.1000000000000001E+0]
        vbroadcastsd    ymm2, qword ptr [rip + .LCPI0_2] # ymm2 = [3.0E+0,3.0E+0,3.0E+0,3.0E+0]
.LBB0_1:                                # =>This Loop Header: Depth=1
        mov     rsi, -448
.LBB0_2:                                #   Parent Loop BB0_1 Depth=1
        vmovupd ymm3, ymmword ptr [rax + rsi - 448]
        vsubpd  ymm3, ymm3, ymmword ptr [rcx + rsi - 448]
        vmovupd ymm4, ymmword ptr [rcx + rsi - 3136]
        vaddpd  ymm4, ymm4, ymm4
        vmulpd  ymm5, ymm0, ymmword ptr [rax + rsi - 3136]
        vaddpd  ymm3, ymm3, ymm4
        vaddpd  ymm3, ymm3, ymm5
        vmovupd ymmword ptr [rdx + rsi - 4928], ymm3
        vmovupd ymm3, ymmword ptr [rax + rsi]
        vsubpd  ymm3, ymm3, ymmword ptr [rcx + rsi]
        vaddpd  ymm3, ymm3, ymmword ptr [rcx + rsi - 1792]
        vmulpd  ymm4, ymm1, ymmword ptr [rax + rsi - 1792]
        vaddpd  ymm3, ymm3, ymm4
        vmulpd  ymm3, ymm3, ymm2
        vmovupd ymmword ptr [rdx + rsi - 2240], ymm3
        vmovupd ymm3, ymmword ptr [rax + rsi + 448]
        vsubpd  ymm3, ymm3, ymmword ptr [rcx + rsi + 448]
        vaddpd  ymm3, ymm3, ymmword ptr [rcx + rsi - 896]
        vmulpd  ymm4, ymm1, ymmword ptr [rax + rsi - 896]
        vaddpd  ymm3, ymm3, ymm4
        vmulpd  ymm3, ymm3, ymm2
        vmovupd ymmword ptr [rdx + rsi + 448], ymm3
        add     rsi, 32
        jne     .LBB0_2
        add     rbx, 1
        add     rax, 4032
        add     rcx, 4032
        add     rdx, 448
        cmp     rbx, 6
        jne     .LBB0_1
        mov     rax, r14
        pop     rbx
        pop     r12
        pop     r13
        pop     r14
        pop     r15
        vzeroupper
        ret
qInterpolatedPlus:
        .quad   0

qInterpolatedMinus:
        .quad   0
```

**Review by Thomas-Ulrich — COMMENTED** [review]
looks good but I let Sebastian validate.

**Comment by ravil-mobile** [comment]
Seems like the compilers struggle to vectorize `postcomputeImposedStateFromNewStress`. It is using `xmm` registers and operate on scalars. Does anybody have any ideas (except `#pragma omp simd`)? 

```
.LCPI0_0:
        .quad   0xc0191eb851eb851f              # double -6.2800000000000002
.LCPI0_1:
        .quad   0x40015c28f5c28f5c              # double 2.1699999999999999
.LCPI0_2:
        .quad   0x4005ae147ae147ae              # double 2.71
.LCPI0_3:
        .quad   0x40091eb851eb851f              # double 3.1400000000000001
postcomputeImposedStateFromNewStress(FaultStresses const&, TractionResults const&, double*, unsigned int): # @postcomputeImposedStateFromNewStress(FaultStresses const&, TractionResults const&, double*, unsigned int)
        push    r14
        push    rbx
        mov     r9, qword ptr [rip + imposedStatePlus]
        mov     r8d, ecx
        mov     r10, qword ptr [rip + imposedStateMinus]
        imul    r14, r8, 3744
        lea     rax, [r9 + r14]
        lea     rcx, [r10 + r14]
        lea     rbx, [r10 + r14]
        add     rbx, 3744
        cmp     rax, rbx
        jae     .LBB0_10
        lea     rbx, [r9 + r14]
        add     rbx, 3744
        cmp     rcx, rbx
        jae     .LBB0_10
        xor     eax, eax
        jmp     .LBB0_3
.LBB0_10:
        mov     ebx, 28
        vxorpd  xmm0, xmm0, xmm0
.LBB0_11:                               # =>This Inner Loop Header: Depth=1
        vmovupd ymmword ptr [rax + 8*rbx - 224], ymm0
        vmovupd ymmword ptr [rax + 8*rbx - 192], ymm0
        vmovupd ymmword ptr [rax + 8*rbx - 160], ymm0
        vmovupd ymmword ptr [rax + 8*rbx - 128], ymm0
        vmovupd ymmword ptr [rcx + 8*rbx - 224], ymm0
        vmovupd ymmword ptr [rcx + 8*rbx - 192], ymm0
        vmovupd ymmword ptr [rcx + 8*rbx - 160], ymm0
        vmovupd ymmword ptr [rcx + 8*rbx - 128], ymm0
        cmp     rbx, 476
        je      .LBB0_12
        vmovupd ymmword ptr [rax + 8*rbx - 96], ymm0
        vmovupd ymmword ptr [rax + 8*rbx - 64], ymm0
        vmovupd ymmword ptr [rax + 8*rbx - 32], ymm0
        vmovupd ymmword ptr [rax + 8*rbx], ymm0
        vmovupd ymmword ptr [rcx + 8*rbx - 96], ymm0
        vmovupd ymmword ptr [rcx + 8*rbx - 64], ymm0
        vmovupd ymmword ptr [rcx + 8*rbx - 32], ymm0
        vmovupd ymmword ptr [rcx + 8*rbx], ymm0
        add     rbx, 32
        jmp     .LBB0_11
.LBB0_12:
        mov     eax, 464
.LBB0_3:
        lea     rcx, [r9 + r14]
        add     rcx, 8
        lea     rbx, [r10 + r14]
        add     rbx, 8
.LBB0_4:                                # =>This Inner Loop Header: Depth=1
        mov     qword ptr [rcx + 8*rax - 8], 0
        mov     qword ptr [rbx + 8*rax - 8], 0
        mov     qword ptr [rcx + 8*rax], 0
        mov     qword ptr [rbx + 8*rax], 0
        add     rax, 2
        cmp     rax, 468
        jne     .LBB0_4
        mov     rax, qword ptr [rip + qInterpolatedPlus]
        mov     rcx, qword ptr [rip + qInterpolatedMinus]
        lea     r11, [r10 + r14]
        add     r11, 3328
        imul    rbx, r8, 22464
        lea     r8, [rcx + rbx]
        add     r8, 3328
        lea     rcx, [r9 + r14]
        add     rcx, 3328
        lea     r9, [rax + rbx]
        add     r9, 3328
        xor     r10d, r10d
        vmovsd  xmm8, qword ptr [rip + .LCPI0_0] # xmm8 = mem[0],zero
        vmovsd  xmm10, qword ptr [rip + .LCPI0_1] # xmm10 = mem[0],zero
        vmovsd  xmm9, qword ptr [rip + .LCPI0_2] # xmm9 = mem[0],zero
        vmovsd  xmm11, qword ptr [rip + .LCPI0_3] # xmm11 = mem[0],zero
.LBB0_6:                                # =>This Loop Header: Depth=1
        vmovsd  xmm4, qword ptr [rdx + 8*r10]   # xmm4 = mem[0],zero
        xor     eax, eax
.LBB0_7:                                #   Parent Loop BB0_6 Depth=1
        vmovsd  xmm7, qword ptr [rdi + rax]     # xmm7 = mem[0],zero
        vmovsd  xmm6, qword ptr [rsi + rax]     # xmm6 = mem[0],zero
        vmovsd  xmm12, qword ptr [rsi + rax + 2496] # xmm12 = mem[0],zero
        vmulsd  xmm0, xmm4, xmm7
        vaddsd  xmm2, xmm0, qword ptr [r11 + rax - 3328]
        vmovsd  qword ptr [r11 + rax - 3328], xmm2
        vmulsd  xmm2, xmm4, xmm6
        vaddsd  xmm1, xmm2, qword ptr [r11 + rax - 2080]
        vmovsd  qword ptr [r11 + rax - 2080], xmm1
        vmulsd  xmm1, xmm12, xmm4
        vaddsd  xmm3, xmm1, qword ptr [r11 + rax - 1248]
        vmovsd  qword ptr [r11 + rax - 1248], xmm3
        vsubsd  xmm3, xmm7, qword ptr [r8 + rax - 3328]
        vmulsd  xmm3, xmm8, xmm3
        vaddsd  xmm3, xmm3, qword ptr [r8 + rax - 832]
        vmulsd  xmm3, xmm4, xmm3
        vaddsd  xmm3, xmm3, qword ptr [r11 + rax - 832]
        vmovsd  qword ptr [r11 + rax - 832], xmm3
        vmovsd  xmm3, qword ptr [r8 + rax - 416] # xmm3 = mem[0],zero
        vsubsd  xmm5, xmm6, qword ptr [r8 + rax - 2080]
        vmulsd  xmm5, xmm10, xmm5
        vsubsd  xmm3, xmm3, xmm5
        vmulsd  xmm3, xmm4, xmm3
        vaddsd  xmm3, xmm3, qword ptr [r11 + rax - 416]
        vmovsd  qword ptr [r11 + rax - 416], xmm3
        vmovsd  xmm3, qword ptr [r8 + rax]      # xmm3 = mem[0],zero
        vsubsd  xmm5, xmm12, qword ptr [r8 + rax - 1248]
        vmulsd  xmm5, xmm10, xmm5
        vsubsd  xmm3, xmm3, xmm5
        vmulsd  xmm3, xmm4, xmm3
        vaddsd  xmm3, xmm3, qword ptr [r11 + rax]
        vmovsd  qword ptr [r11 + rax], xmm3
        vaddsd  xmm0, xmm0, qword ptr [rcx + rax - 3328]
        vmovsd  qword ptr [rcx + rax - 3328], xmm0
        vaddsd  xmm0, xmm2, qword ptr [rcx + rax - 2080]
        vmovsd  qword ptr [rcx + rax - 2080], xmm0
        vaddsd  xmm0, xmm1, qword ptr [rcx + rax - 1248]
        vmovsd  qword ptr [rcx + rax - 1248], xmm0
        vsubsd  xmm0, xmm7, qword ptr [r9 + rax - 3328]
        vmulsd  xmm0, xmm9, xmm0
        vaddsd  xmm0, xmm0, qword ptr [r9 + rax - 832]
        vmulsd  xmm0, xmm4, xmm0
        vaddsd  xmm0, xmm0, qword ptr [rcx + rax - 832]
        vmovsd  qword ptr [rcx + rax - 832], xmm0
        vsubsd  xmm0, xmm6, qword ptr [r9 + rax - 2080]
        vmulsd  xmm0, xmm11, xmm0
        vaddsd  xmm0, xmm0, qword ptr [r9 + rax - 416]
        vmulsd  xmm0, xmm4, xmm0
        vaddsd  xmm0, xmm0, qword ptr [rcx + rax - 416]
        vmovsd  qword ptr [rcx + rax - 416], xmm0
        vsubsd  xmm0, xmm12, qword ptr [r9 + rax - 1248]
        vmulsd  xmm0, xmm11, xmm0
        vaddsd  xmm0, xmm0, qword ptr [r9 + rax]
        vmulsd  xmm0, xmm4, xmm0
        vaddsd  xmm0, xmm0, qword ptr [rcx + rax]
        vmovsd  qword ptr [rcx + rax], xmm0
        add     rax, 8
        cmp     rax, 416
        jne     .LBB0_7
        add     r10, 1
        add     rdi, 416
        add     rsi, 416
        add     r8, 3744
        add     r9, 3744
        cmp     r10, 6
        jne     .LBB0_6
        pop     rbx
        pop     r14
        vzeroupper
        ret
qInterpolatedPlus:
        .quad   0

qInterpolatedMinus:
        .quad   0

imposedStatePlus:
        .quad   0

imposedStateMinus:
        .quad   0
```

**Comment by ravil-mobile** [comment]
I just made put `postcomputeImposedStateFromNewStress` to a separate cpp file and tried to compile like:
```
g++ ./vec.cpp -c -O3 -DNDEBUG -mavx2 -mfma -ftree-vectorize -fopt-info-vec-missed 
```

here is the info:

```
./vec.cpp:53:28: missed: couldn't vectorize loop
./vec.cpp:75:79: missed: not vectorized: complicated access pattern.
./vec.cpp:56:30: missed: couldn't vectorize loop
./vec.cpp:56:30: missed: number of versioning for alias run-time tests exceeds 10 (--param vect-max-version-for-alias-checks)
./vec.cpp:24:6: missed: statement clobbers memory: __builtin_memset (_4, 0, 3744);
./vec.cpp:24:6: missed: statement clobbers memory: __builtin_memset (_8, 0, 3744);
```

**Comment by Thomas-Ulrich** [comment]
Ravil,  why do you think it will be faster when moved back to c++ from yateto?

**Comment by ravil-mobile** [comment]
>Also, isn't it a bit early to do optimizations?

Yes, you right. I just wanted to try

**Comment by Thomas-Ulrich** [comment]
Ok, I don't have the background to be able to judge if it will be faster or not, but your reasoning makes sense (more portable and maintainable for the same performance).

**Comment by krenzland** [comment]
I mean, if you want to force SIMD, the easiest way is probably to define a a struct vectorizedInt and implement add/mul operators on them using intrinsics. This way, you have to define the struct for each CPU arch but can reuse more or less the same implementation.
That's a pattern that I've seen in other codes and it seems to work quite well in practice.

**Comment by krenzland** [comment]
I mean, if you want to force SIMD, the easiest way is probably to define a a struct vectorizedInt and implement add/mul operators on them using intrinsics. This way, you have to define the struct for each CPU arch but can reuse more or less the same implementation.
That's a pattern that I've seen in other codes and it seems to work quite well in practice.

**Comment by ravil-mobile** [comment]
> I mean, if you want to force SIMD, the easiest way is probably to define a a struct vectorizedInt and implement add/mul operators on them using intrinsics. This way, you have to define the struct for each CPU arch but can reuse more or less the same implementation. That's a pattern that I've seen in other codes and it seems to work quite well in practice.

I just wanted to express the code in such a way that `gcc` and `clang` could vectorize it without any hints. `pragma omp simd` nicely works for `postcomputeImposedStateFromNewStress`. I tried `compiler guided vectorization` from `icpc`. I understood that vectorization is prevented because of potential memory aliases. I think the code is lacking `restrict` key words somewhere.

I think `pragma omp simd`  is ok.

**Review by sebwolf-de — APPROVED** [review]
All in all, these changes make the code clearer to read. When I'll be working on Dynamic Rupture with poroelastic materials, I have to touch this part again, and I am happy that it is easier now.
If this version is faster or the yateto version, is up to benchmarking. Yet, at this point, I prefer this version.
Thanks @ravil-mobile for taking care of this and also for checking the vectorization details.

**Comment by krenzland** [comment]
> I think the code is lacking `restrict` key words somewhere.
Yup, probably. Afaik pragma omp simd doesn't actually do much except to say "reorder/vectorize this, it's gonna work - I promise"

File: src/DynamicRupture/FrictionLaws/BaseFrictionLaw.h [hunk 1]
```diff
@@ -89,36 +89,41 @@ class BaseFrictionLaw : public FrictionSolver {
...
-      stressFromQInterpolatedKrnl.execute();
+
+    auto etaP = impAndEta[ltsFace].etaP;
+    auto etaS = impAndEta[ltsFace].etaS;
+    auto invZp = impAndEta[ltsFace].invZp;
+    auto invZs = impAndEta[ltsFace].invZs;
+    auto invZpNeig = impAndEta[ltsFace].invZpNeig;
+    auto invZsNeig = impAndEta[ltsFace].invZsNeig;
+
+    constexpr int numQuantities{9};
```
Comment (Thomas-Ulrich) [inline src/DynamicRupture/FrictionLaws/BaseFrictionLaw.h:107]: does this work with attenuation?

**Pull request merged**